File: robocode2-bot-api-java/src/main/java/net/robocode2/mapper/EventMapper.java
package net.robocode2.mapper;

import lombok.NonNull;
import lombok.experimental.UtilityClass;
import lombok.val;
import net.robocode2.BotException;
import net.robocode2.events.*;

import java.util.ArrayList;
import java.util.List;

/** Utility class for mapping events */
@UtilityClass
public class EventMapper {

  public TickEvent map(@NonNull final net.robocode2.schema.TickEventForBot source) {
    return TickEvent.builder()
        .turnNumber(source.getTurnNumber())
        .roundNumber(source.getRoundNumber())
        .botState(BotStateMapper.map(source.getBotState()))
        .bulletStates(BulletStateMapper.map(source.getBulletStates()))
        .events(map(source.getEvents()))
        .build();
  }

  private List<GameEvent> map(@NonNull final List<net.robocode2.schema.Event> source) {
    val gameEvents = new ArrayList<GameEvent>();
    source.forEach(event -> gameEvents.add(map(event)));
    return gameEvents;
  }

  private GameEvent map(@NonNull final net.robocode2.schema.Event source) {
    if (source instanceof net.robocode2.schema.BotDeathEvent) {
      return map((net.robocode2.schema.BotDeathEvent) source);
    }
    if (source instanceof net.robocode2.schema.BotHitBotEvent) {
      return map((net.robocode2.schema.BotHitBotEvent) source);
    }
    if (source instanceof net.robocode2.schema.BotHitWallEvent) {
      return map((net.robocode2.schema.BotHitWallEvent) source);
    }
    if (source instanceof net.robocode2.schema.BulletFiredEvent) {
      return map((net.robocode2.schema.BulletFiredEvent) source);
    }
    if (source instanceof net.robocode2.schema.BulletHitBotEvent) {
      return map((net.robocode2.schema.BulletHitBotEvent) source);
    }
    if (source instanceof net.robocode2.schema.BulletHitBulletEvent) {
      return map((net.robocode2.schema.BulletHitBulletEvent) source);
    }
    if (source instanceof net.robocode2.schema.BulletHitWallEvent) {
      return map((net.robocode2.schema.BulletHitWallEvent) source);
    }
    if (source instanceof net.robocode2.schema.ScannedBotEvent) {
      return map((net.robocode2.schema.ScannedBotEvent) source);
    }
    if (source instanceof net.robocode2.schema.SkippedTurnEvent) {
      return map((net.robocode2.schema.SkippedTurnEvent) source);
    }
    if (source instanceof net.robocode2.schema.WonRoundEvent) {
      return map((net.robocode2.schema.WonRoundEvent) source);
    }
    throw new BotException(
        "No mapping exists for event type: " + source.getClass().getSimpleName());
  }

  private BotDeathEvent map(@NonNull final net.robocode2.schema.BotDeathEvent source) {
    return BotDeathEvent.builder()
        .turnNumber(source.getTurnNumber())
        .victimId(source.getVictimId())
        .build();
  }

  private BotHitBotEvent map(@NonNull final net.robocode2.schema.BotHitBotEvent source) {
    return BotHitBotEvent.builder()
        .turnNumber(source.getTurnNumber())
        .botId(source.getBotId())
        .victimId(source.getVictimId())
        .energy(source.getEnergy())
        .x(source.getX())
        .y(source.getY())
        .rammed(source.getRammed())
        .build();
  }

  private BotHitWallEvent map(@NonNull final net.robocode2.schema.BotHitWallEvent source) {
    return BotHitWallEvent.builder()
        .turnNumber(source.getTurnNumber())
        .victimId(source.getVictimId())
        .build();
  }

  private BulletFiredEvent map(@NonNull final net.robocode2.schema.BulletFiredEvent source) {
    return BulletFiredEvent.builder()
        .turnNumber(source.getTurnNumber())
        .bullet(BulletStateMapper.map(source.getBullet()))
        .build();
  }

  private BulletHitBotEvent map(@NonNull final net.robocode2.schema.BulletHitBotEvent source) {
    return BulletHitBotEvent.builder()
        .turnNumber(source.getTurnNumber())
        .victimId(source.getVictimId())
        .bullet(BulletStateMapper.map(source.getBullet()))
        .damage((source.getDamage()))
        .energy(source.getEnergy())
        .build();
  }

  private BulletHitBulletEvent map(
      @NonNull final net.robocode2.schema.BulletHitBulletEvent source) {
    return BulletHitBulletEvent.builder()
        .turnNumber(source.getTurnNumber())
        .bullet(BulletStateMapper.map(source.getBullet()))
        .hitBullet(BulletStateMapper.map(source.getHitBullet()))
        .build();
  }

  private BulletHitWallEvent map(@NonNull final net.robocode2.schema.BulletHitWallEvent source) {
    return BulletHitWallEvent.builder()
        .turnNumber(source.getTurnNumber())
        .bullet(BulletStateMapper.map(source.getBullet()))
        .build();
  }

  private ScannedBotEvent map(@NonNull final net.robocode2.schema.ScannedBotEvent source) {
    return ScannedBotEvent.builder()
        .turnNumber(source.getTurnNumber())
        .scannedByBotId(source.getScannedByBotId())
        .scannedBotId(source.getScannedBotId())
        .energy(source.getEnergy())
        .x(source.getX())
        .y(source.getY())
        .direction((source.getDirection()))
        .speed(source.getSpeed())
        .build();
  }

  private SkippedTurnEvent map(@NonNull final net.robocode2.schema.SkippedTurnEvent source) {
    return SkippedTurnEvent.builder().turnNumber(source.getTurnNumber()).build();
  }

  private WonRoundEvent map(@NonNull final net.robocode2.schema.WonRoundEvent source) {
    return WonRoundEvent.builder().turnNumber(source.getTurnNumber()).build();
  }
}

File: robocode-tankroyale-bot-api-csharp/src/mapper/EventMapper.cs
using System.Collections.Generic;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using Robocode.TankRoyale.BotApi.Events;

namespace Robocode.TankRoyale.BotApi
{
  public sealed class EventMapper
  {
    public static TickEvent Map(string json)
    {
      var source = JsonConvert.DeserializeObject<Schema.TickEventForBot>(json);

      var jsonTickEvent = JsonConvert.DeserializeObject<Dictionary<string, object>>(json);
      JArray events = (JArray)jsonTickEvent["events"];

      return new TickEvent(
        source.TurnNumber,
        source.RoundNumber,
        source.EnemyCount,
        BotStateMapper.Map(source.BotState),
        BulletStateMapper.Map(source.BulletStates),
        Map(events)
      );
    }

    private static HashSet<BotEvent> Map(JArray events)
    {
      var gameEvents = new HashSet<BotEvent>();
      foreach (JObject evt in events)
      {
        gameEvents.Add(Map(evt));
      }
      return gameEvents;
    }

    public static BotEvent Map(JObject evt)
    {
      string type = evt.GetValue("$type").ToString();

      switch (type)
      {
        case "DeathEvent":
          return Map(evt.ToObject<Schema.BotDeathEvent>());
        case "BotHitBotEvent":
          return Map(evt.ToObject<Schema.BotHitBotEvent>());
        case "BotHitWallEvent":
          return Map(evt.ToObject<Schema.BotHitWallEvent>());
        case "BulletFiredEvent":
          return Map(evt.ToObject<Schema.BulletFiredEvent>());
        case "BulletHitBotEvent":
          return Map(evt.ToObject<Schema.BulletHitBotEvent>());
        case "BulletHitBulletEvent":
          return Map(evt.ToObject<Schema.BulletHitBulletEvent>());
        case "BulletHitWallEvent":
          return Map(evt.ToObject<Schema.BulletHitWallEvent>());
        case "ScannedBotEvent":
          return Map(evt.ToObject<Schema.ScannedBotEvent>());
        case "SkippedTurnEvent":
          return Map(evt.ToObject<Schema.SkippedTurnEvent>());
        case "WonRoundEvent":
          return Map(evt.ToObject<Schema.WonRoundEvent>());
        default:
          throw new BotException("No mapping exists for event type: " + type);
      }
    }

    private static DeathEvent Map(Schema.BotDeathEvent source)
    {
      return new DeathEvent(
        source.TurnNumber,
        source.VictimId
      );
    }

    private static HitBotEvent Map(Schema.BotHitBotEvent source)
    {
      return new HitBotEvent(
        source.TurnNumber,
        source.VictimId,
        source.Energy,
        source.X,
        source.Y,
        source.Rammed
      );
    }

    private static HitWallEvent Map(Schema.BotHitWallEvent source)
    {
      return new HitWallEvent(source.TurnNumber);
    }

    private static BulletFiredEvent Map(Schema.BulletFiredEvent source)
    {
      return new BulletFiredEvent(
        source.TurnNumber,
        BulletStateMapper.Map(source.Bullet)
      );
    }

    private static BulletHitBotEvent Map(Schema.BulletHitBotEvent source)
    {
      return new BulletHitBotEvent(
        source.TurnNumber,
        source.VictimId,
        BulletStateMapper.Map(source.Bullet),
        source.Damage,
        source.Energy
      );
    }

    private static BulletHitBulletEvent Map(Schema.BulletHitBulletEvent source)
    {
      return new BulletHitBulletEvent(
        source.TurnNumber,
        BulletStateMapper.Map(source.Bullet),
        BulletStateMapper.Map(source.HitBullet)
      );
    }

    private static BulletHitWallEvent Map(Schema.BulletHitWallEvent source)
    {
      return new BulletHitWallEvent(
        source.TurnNumber,
        BulletStateMapper.Map(source.Bullet)
      );
    }

    private static ScannedBotEvent Map(Schema.ScannedBotEvent source)
    {
      return new ScannedBotEvent(
        source.TurnNumber,
        source.ScannedByBotId,
        source.ScannedBotId,
        source.Energy,
        source.X,
        source.Y,
        source.Direction,
        source.Speed
      );
    }

    public static SkippedTurnEvent Map(Schema.SkippedTurnEvent source)
    {
      return new SkippedTurnEvent(
        source.TurnNumber
      );
    }

    private static WonRoundEvent Map(Schema.WonRoundEvent source)
    {
      return new WonRoundEvent(
        source.TurnNumber
      );
    }
  }
}